python/src/ignition/gazebo/world.cc

```diff
@@ -0,0 +1,75 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <pybind11/pybind11.h>
+
+#include <iostream>
+
+#include "world.hh"
+
+namespace ignition
+{
+namespace gazebo
+{
+namespace python
+{
+World::World(ignition::gazebo::Entity _entity)
+{
+  _world = std::make_shared<ignition::gazebo::World>(_entity);
+}
+
+World::~World()
+{
+}
+
+void World::destroy()
+{
+  _world.reset();
+}
+
+ignition::math::Vector3<double> World::Gravity(const EntityComponentManager &_ecm)
+{
+  std::optional<ignition::math::Vector3d> gravity = _world->Gravity(*_ecm.rcl_ptr());
+  if (gravity.has_value())
+  {
+    return gravity.value();
+  }
+  return ignition::math::Vector3d::NaN;
+}
+
+ignition::gazebo::Entity World::ModelByName(
+  ignition::gazebo::python::EntityComponentManager &_ecm,
+  std::string _name)
+{
+  return _world->ModelByName(*_ecm.rcl_ptr(), _name);
+}
+
+void define_gazebo_world(py::object module)
+{
+  py::class_<ignition::gazebo::python::World,
+             ignition::utils::python::Destroyable,
+             std::shared_ptr<ignition::gazebo::python::World>>(module, "World")
+  .def(py::init<ignition::gazebo::Entity>())
+  .def(
+    "model_by_name", &ignition::gazebo::python::World::ModelByName,
```

Comment:
It's not clear to me why we need to repeat the functions in the `python` namespace. What's preventing us from binding to `ignition::gazebo::World::ModelByName` here?

Reply:
`std::optional` can't be used as a returned type in methods which means we need to create a class to workaround this, then the rest of the methods should be wrapped too, beacuse I created a `World` instance.

Reply:
Ouch, I thought pybind11 had support for C++17, including `std::optional`, and that was one of its advantages with respect to SWIG?

https://pybind11.readthedocs.io/en/stable/advanced/cast/stl.html#c-17-library-containers

_The pybind11/stl.h header also includes support for std::optional<> and std::variant<>._

Reply:
https://github.com/ignitionrobotics/ign-gazebo/pull/1219/commits/dbe9c08e435079c6d4ca47d8980a2be07f47e326

Reply:
So now that we're using optional, do we still need to redefine these functions?

Reply:
every class that make use of the EntityComponentManager should be wrap

Reply:
Oh that's going to be hard to scale. Is there any change we can make to the original `EntityComponentManager` class so we don't need to wrap it? Maybe we just need to make the original class copiable?